### PR TITLE
Add support for basic auth

### PR DESCRIPTION
**Pull request by unstubbable**
~Note: When we have agreed on an API, I will update the README as well.~ ✅

### CODE REVIEW

File: src/cdk/utils/basic-authorizer-handler/index.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  return {
+    principalId,
+    policyDocument: {
+      Version: '2012-10-17',
+      Statement: [
+        {
+          Action: 'execute-api:Invoke',
+          Effect: 'Allow',
+          Resource:
+            resource
```
Comment (clebert) [inline src/cdk/utils/basic-authorizer-handler/index.ts]: Maybe add an example comment before vs after

File: src/cdk/utils/create-unauthorized-gateway-response.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      'gatewayresponse.header.WWW-Authenticate': realm
+        ? `'Basic realm="${realm}"'`
+        : "'Basic'"
+    },
+    responseTemplates: {
+      'application/json': '{"message":$context.error.messageString}'
+    }
+  });
+
+  // Intentionally not adding the restApi as a dependency here, otherwise the
```
Comment (clebert) [inline src/cdk/utils/create-unauthorized-gateway-response.ts]: Please be more verbose

File: src/cdk/utils/create-unauthorized-gateway-response.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  // We need to use a low-level construct here that should be replaced when
+  // @aws-cdk/aws-apigateway adds a high-level construct for gateway responses.
+  new CfnGatewayResponse(stack, 'ApiGatewayUnauthorizedResponse', {
+    statusCode: '401',
+    responseType: 'UNAUTHORIZED',
+    restApiId: restApi.restApiId,
+    responseParameters: {
+      'gatewayresponse.header.WWW-Authenticate': realm
+        ? `'Basic realm="${realm}"'`
```
Comment (clebert) [inline src/cdk/utils/create-unauthorized-gateway-response.ts]: Please remove the realm feature

File: src/types.ts [hunk 4]
```diff
@@ -82,6 +91,7 @@ export interface StackConfig {
   readonly minimumCompressionSizeInBytes?: number;
   readonly lambdaConfigs?: LambdaConfig[];
   readonly s3Configs?: S3Config[];
+  readonly basicAuthentication?: BasicAuthenticationConfig;
```
Comment (clebert) [inline src/types.ts]: Maybe add "Quercheck" with good error message in aws-simple

Comment (clebert) [inline src/types.ts]: basicAuthenticationConfig

File: src/cdk/utils/create-unauthorized-gateway-response.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        : "'Basic'"
+    },
+    responseTemplates: {
+      'application/json': '{"message":$context.error.messageString}'
```
Comment (clebert) [inline src/cdk/utils/create-unauthorized-gateway-response.ts]: Add text content type response

File: src/cdk/utils/create-lambda-integration.ts [hunk 6]
```diff
@@ -52,6 +63,10 @@ export function createLambdaIntegration(
       }
     ),
     {
+      authorizationType: authenticationRequired
+        ? AuthorizationType.CUSTOM
+        : AuthorizationType.NONE,
+      authorizer: authenticationRequired ? undefined : noneAuthorizer,
```
Comment (clebert) [inline src/cdk/utils/create-lambda-integration.ts]: Pass authorizer into this function, and throw if authenticationRequired is true but no authorizer is passed

File: src/cdk/utils/create-rest-api-props.ts [hunk 7]
```diff
@@ -16,6 +17,9 @@ export function createRestApiProps(
     domainName: createDomainNameOptions(stackConfig, stack),
     binaryMediaTypes,
     minimumCompressionSize: minimumCompressionSizeInBytes,
-    deployOptions: createStageOptions(stackConfig)
+    deployOptions: createStageOptions(stackConfig),
+    defaultMethodOptions: {
```
Comment (clebert) [inline src/cdk/utils/create-rest-api-props.ts]: remove

File: src/cdk/utils/create-lambda-integration.ts [hunk 8]
```diff
@@ -10,15 +10,11 @@ import * as path from 'path';
...
-  authorizerId: 'none',
-  authorizationType: AuthorizationType.NONE
-};
-
 export function createLambdaIntegration(
   stack: Stack,
   restApi: RestApi,
-  lambdaConfig: LambdaConfig
+  lambdaConfig: LambdaConfig,
+  authorizer?: IAuthorizer
```
Comment (clebert) [inline src/cdk/utils/create-lambda-integration.ts]: I thin this should be a non-optional parameter.

File: src/cdk/utils/create-s3-integration.ts [hunk 9]
```diff
@@ -11,16 +11,12 @@ import {S3Config} from '../../types';
...
-  authorizationType: AuthorizationType.NONE
-};
-
 export function createS3Integration(
   restApi: RestApi,
   s3Bucket: Bucket,
   s3IntegrationRole: Role,
-  s3Config: S3Config
+  s3Config: S3Config,
+  authorizer?: IAuthorizer
```
Comment (clebert) [inline src/cdk/utils/create-s3-integration.ts]: Same here

File: src/cdk/utils/create-unauthorized-gateway-response.ts [hunk 10]
```diff
@@ -7,30 +7,31 @@ export function createUnauthorizedGatewayResponse(
   stack: Stack,
   restApi: RestApi
 ): void {
-  const {basicAuthentication} = stackConfig;
+  const {basicAuthenticationConfig: basicAuthentication} = stackConfig;
```
Comment (clebert) [inline src/cdk/utils/create-unauthorized-gateway-response.ts]: Why the renaming? I think we can use the property directly without destructuring

File: src/cdk/utils/create-unauthorized-gateway-response.ts (hunk 5 above)
Comment (clebert) [reply src/cdk/utils/create-unauthorized-gateway-response.ts]: Do you thought about that?

File: src/cdk/utils/create-unauthorized-gateway-response.ts (hunk 10 above)
Comment (unstubbable) [reply src/cdk/utils/create-unauthorized-gateway-response.ts]: accidental via F2

File: src/cdk/utils/create-lambda-integration.ts (hunk 8 above)
Comment (unstubbable) [reply src/cdk/utils/create-lambda-integration.ts]: Agree

File: src/cdk/utils/create-unauthorized-gateway-response.ts (hunk 5 above)
Comment (unstubbable) [reply src/cdk/utils/create-unauthorized-gateway-response.ts]: Did you see it?

Comment (clebert) [reply src/cdk/utils/create-unauthorized-gateway-response.ts]: Yes my view was outdated

**Pull request merged**